app/javascript/src/PaymentGateways/braintree/BraintreeForm.jsx

```diff
@@ -54,12 +56,20 @@ const BraintreeForm = ({
     }
   }, [braintreeNonceValue])
 
+  const get3DSecureError = (response) => {
+    return (
+      response.name === 'BraintreeError' ||
+        !!response.threeDSecureInfo.status.match(/authenticate_(attempt_)?successful/) === false
```

Comment:
I don't know about js, but are there any benefits of using `!!response.threeDSecureInfo.status.match(/authenticate_(attempt_)?successful/) === false` instead of `!response.threeDSecureInfo.status.match(/authenticate_(attempt_)?successful/)`?

Reply:
@akostadinov the double `!!` will cast `true` or `false` out of every value, `null`, `undefined` or `0` would be `false` and pretty much any other `string`, `number`, `object` or `array`(even if they are empty) would be `true`

Reply:
@akostadinov You are right, even tho the double bang will cast every "falsey" value to `false`, the single one will turn that "truthy" one matching our assertion

Reply:
@akostadinov @didierofrivia  

The problem is not when _falsy_, but when _truthy_.
If `response.threeDSecureInfo.status.match(/authenticate_(attempt_)?successful/)` finds a match, it will return an array, then it will be flipped to `false`, thanks to JS dynamic type coercion:

<img width="382" alt="Screenshot 2021-08-18 at 10 18 35" src="https://user-images.githubusercontent.com/13486237/129863479-ba1c532d-5bae-4aa5-ac38-477b1e0ac471.png">

That's why we need the longest:
`!!response.threeDSecureInfo.status.match(/authenticate_(attempt_)?successful/) === false`

Reply:
In the other hand, yes, we should check if `response.threeDSecureInfo.status` is defined, I will push the change

Reply:
Yet another reason to have good types and understanding of their API. A week from now and we will forget why the double band and false assertion...

Reply:
We should create a new task for that, adding types and improving error messages

Reply:
@akostadinov ⬆️